### PR TITLE
First pass at "Objectifying" Mongo classes.

### PR DESCRIPTION
**Pull request by larspetrus**

### CODE REVIEW

File: both/_round.js [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+_.extend(Round.prototype, {
+
+  format: function() {
+    return wca.formatByCode[this.formatCode];
+  },
+  properties: function() {
+    return wca.roundByCode[this.roundCode];
+  },
+  eventName: function() {
+    return this.eventCode ? wca.eventByCode[this.eventCode].name : '';
```
Comment (jfly) [inline both/_round.js]: Why the ternary? Is this to handle non wca Rounds, or excluded fields? Either way, I think we should assert. (same for eventName).

File: both/_round.js [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+  },
+  eventSolveTimeFields: function() {
```
Comment (jfly) [inline both/_round.js]: Since we've got eventName and eventSolveTimeFields, could we just have a single method that returns the event instead?

File: both/_round.js [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+    return this.status === wca.roundStatuses.open;
+  },
+  isClosed: function() {
+    return this.status === wca.roundStatuses.closed;
+  },
+});
+
+
+
+MIN_ROUND_DURATION_MINUTES = 30;
```
Comment (jfly) [inline both/_round.js]: Unrelated cleaup idea: could we make these static attributes of Round instead?

File: both/helpers.js [hunk 4]
```diff
@@ -1,322 +1,296 @@
-if(Meteor.isClient) {
-  Template.registerHelper("roundName", function(roundCode) {
+(function() { // This is a Closure
```
Comment (jfly) [inline both/helpers.js]: Why the closure? Meteor already does this automatically. See http://staging.live.jflei.com/both/helpers.js?ba980f78122081392373894bb73156b9143d5c11.

File: both/methods.js [hunk 5]
```diff
@@ -248,35 +247,30 @@ Meteor.methods({
       Groups.insert(newGroup);
     }
   },
-  advanceParticipantsFromRound: function(participantCount, roundId) {
-    var competitionId = getRoundAttribute(roundId, 'competitionId');
-    throwIfCannotManageCompetition(this.userId, competitionId);
+  advanceParticipantsFromRound: function(participantsToAdvance, roundId) {
```
Comment (jfly) [inline both/methods.js]: Good argument rename!

File: tests/mocha/server/RoundTest.js [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+    it('eventSolveTimeFields()', function() {
+      chai.expect(makeRound({ eventCode: '333mbf' }).eventSolveTimeFields().join()).to.equal("millis,puzzlesSolvedCount,puzzlesAttemptedCount");
+      chai.expect(lunch.eventSolveTimeFields()).to.be.undefined;
+    });
+
+    it('status convenience functions', function() {
+      var unstarted = makeRound({ status: 'unstarted' });
+      var open      = makeRound({ status: 'open' });
+      var closed    = makeRound({ status: 'closed' });
```
Comment (jfly) [inline tests/mocha/server/RoundTest.js]: I do not like aligning equals signs. I find that it means you have to edit lots of lines at some point, or your code becomes inconsistent.

File: tests/mocha/server/RoundTest.js [hunk 7]
```diff
@@ -0,0 +1,52 @@
+MochaWeb.testOnly(function() {
```
Comment (jfly) [inline tests/mocha/server/RoundTest.js]: This is awesome! So cool to have tests of actual meteor code =)

File: both/_round.js [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+// Add functions to the Mongo object, using transform (see http://docs.meteor.com/#/full/mongo_collection)
+Round = function(doc) {
+  _.extend(this, doc);
+};
+_.extend(Round.prototype, {
+
+  format: function() {
+    return wca.formatByCode[this.formatCode];
+  },
+  properties: function() {
```
Comment (jfly) [inline both/_round.js]: Hmm... I've never liked our "format" vs "round" name. roundFormat vs roundProperties seems better (still not in love with it, though). We don't need to do this as part of this PR, but I would like to make the naming inside of `__wca.js` consistent with this.

**Comment by jfly** [comment]
Thanks, Lars! This is an all around improvement. My biggest concern is that there are a bunch of places where we are now looking up an entire Round, rather than just the attribute we need from that round. This wouldn't be a problem, as most of the attribute of a Round don't change that often (people don't reschedule rounds often, for instance). However, we have this progress field https://github.com/cubing/ccm/blob/master/both/_collections.js#L555 that is going to change very often (every time a solve is entered). Urg. My understanding is that the "best practice" is to always only ask minimongo for just the fields you need. That way you won't run into potentially serious reactivity related performance issues. I know it's annoying to do, but if you always do it, you never have to think about it.

File: both/_round.js (hunk 1 above)
Comment (larspetrus) [inline both/_round.js]: For non competition rounds. Which BTW I think should be their own table/class, but that's not for now.

I don't understand asserts in production code, but I can add whatever you think should be there.

File: both/_round.js (hunk 2 above)
Comment (larspetrus) [inline both/_round.js]: Maybe. This API seemed nicer to me. There are only two fields, and they seem unrelated, while the others look more like coherent structures.

But that just me winging it. I have neither a strong opinion or a deep understanding.

File: both/helpers.js (hunk 4 above)
Comment (larspetrus) [inline both/helpers.js]: Oh dear...

Here is what happened:
Through a RubyMine mishap, I did this editing  in a generated helpers.js file under .meteor/. When I discovered that, I didn't want to redo all the work, so I copied the whole file into the regular file.

Then I noticed the closure, which I didn't understand, but didn't want to remove it. I added the comment, since I found it confusing, and also I was proud to have spotted it.

File: both/_round.js (hunk 1 above)
Comment (jfly) [inline both/_round.js]: Yeah, moving them to their own table seems like a reasonable thing to do.

If this method shouldn't be called, then I'd rather it barf instead of silently returning something. Lets `assert(this.eventCode)`.

File: both/_round.js (hunk 2 above)
Comment (jfly) [inline both/_round.js]: I see. Lets leave it as is.

File: both/helpers.js (hunk 4 above)
Comment (jfly) [inline both/helpers.js]: Ahh, that explains where it came from. Lets remove it =).

File: tests/mocha/server/RoundTest.js (hunk 6 above)
Comment (larspetrus) [inline tests/mocha/server/RoundTest.js]: Aligning code is banned at Google. It's derisively called "ASCII art". To me, increased readability always wins over some extra writing effort, but I know that's a minority view and I'll change this to the project standard.

Comment (jfly) [inline tests/mocha/server/RoundTest.js]: That's interesting. Thanks for deferring =)

Comment (FatBoyXPC) [inline tests/mocha/server/RoundTest.js]: I'm pretty sure I count for at least 2 votes due to size and I love aligning equals :-)

File: both/_round.js (hunk 8 above)
Comment (larspetrus) [inline both/_round.js]: I just couldn't have "round.round" in the code, so came up with this bland vague name.

I think it's hard to name these things because they should really be individual properties or functions on the Round class itself. "code" is already there, and the others look quite computable.

File: both/_round.js (hunk 3 above)
Comment (larspetrus) [inline both/_round.js]: Could and should. I would expose MIN_ROUND_DURATION_MINUTES &
DEFAULT_ROUND_DURATION_MINUTES, and inline the unused DEFAULT_ROUND_NTHDAY.

File: both/_round.js (hunk 8 above)
Comment (jfly) [inline both/_round.js]: Yeah, the extra layer of heirarchy here does seem pretty arbitrary. Something to improve another day =)

File: both/_round.js (hunk 3 above)
Comment (jfly) [inline both/_round.js]: Wow, did not realize DEFAULT_ROUND_NTHDAY is unused, inline sounds good!

**Comment by larspetrus** [comment]
When we discussed this earlier you seemed open to loading full objects, combined with separating data that could interact badly into separate tables.

In this case I think putting the progress data in its own table makes a lot of sense. Both from the performance perspective you mention, and since it's summarized Result data rather than actual properties of a Round.

I'm strongly in favor of always loading whole objects, so I guess we should talk this out some time. But I don't think this PR depends on reaching philosophical alignment.

**Comment by jfly** [comment]
Sounds good. Let's move forward with this, and if performance issues come up later on, we'll address them. 

Moving progress to a separate table seems like a reasonable compromise.

**Comment by jfly** [comment]
And apologies for flip flopping. I hadn't anticipated the progress field, upon which I was basing most of my previous point. People aren't going to change stuff like the start time of a round often, so there's no need to optimize for that case.

File: both/_round.js (hunk 1 above)
Comment (larspetrus) [inline both/_round.js]: Yeah, but I would use a regular exception if barfing is needed.

Anyway, I put in the asserts. Once we break out the ScheduleItem class, eventCode should become mandatory, and this problem disappears in a puff of good design.

**Comment by larspetrus** [comment]
OK, I'll break out the RoundProgress class/table next. And split up _collections.js. After sleeping on it, how I left it now looks disgraceful.

**Comment by jfly** [comment]
These all sound like great plans! If at all possible, try to split up into multiple PRs, it makes it easier to review =). Is this one ready to merge?
